transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -311,19 +311,19 @@ private void processReady(EpollEventArray events, int ready) {
 
                     AbstractEpollUnsafe unsafe = (AbstractEpollUnsafe) ch.unsafe();
 
-                    if (close) {
-                        unsafe.epollRdHupReady();
-                    }
-
                     // We need to check if the channel is still open before try to trigger the
-                    // callbacks as otherwise we may trigger an IllegalStateException when try
-                    // to access the file descriptor.
+                    // callbacks.
                     //
                     // See https://github.com/netty/netty/issues/3443
+                    if (close && ch.isOpen()) {
+                        unsafe.epollRdHupReady();
+                    }
+
                     if (write && ch.isOpen()) {
```

Comment:
Consider inlining `write` (and `close`/`read`).  They are only used 1 time...I imagine this method is called frequently too.